### ISSUE TITLE
Spark Out of memory during fitpathsmultidataset

### ISSUE THREAD

**Issue by kris1710** [deeplearning4j/deeplearning4j#8467]
hi can anyone suggest any workaround for this. I am running a computation graph on spark. I prepared my multidataset objects (input size =140 features,batch size 32) and saved the paths in the hdfs. There are total 290K multidataset objects.

```
val trainingMaster: ParameterAveragingTrainingMaster = new ParameterAveragingTrainingMaster.Builder(60,32).workerPrefetchNumBatches(0).averagingFrequency(4).batchSizePerWorker(32).build();
val sparknet=new SparkComputationGraph(sc,net5,trainingMaster)

val data_paths="hdfs:////user/dh-ads/karun/proc1.txt"
val loadedPaths = sc.textFile(data_paths)
sparknet.fitMultiDataSet(loadedPaths)

```
The job fails after running for sometime

My model has 52 million parameters to learn. I call net.fitMultidatasetPaths(<path to my hdfs directory>) however, it fails during the training giving the following error

```
Container killed for exceeding the memory limits
```
I used ```trainingWorkspaceMode(WorkspaceMode.ENABLED)```  while defining the neural net configuration.

**Comment by AlexDBlack** [comment]
Take a look at this: https://deeplearning4j.org/docs/latest/deeplearning4j-scaleout-howto#memory
My guess is that you're running on YARN but haven't configured memory overhead and/or ND4J memory settings?
The default memory overhead (i.e., amount of off-heap/non-JVM memory allowed in the container) is only like 1GB.

**Comment by kris1710** [comment]
@AlexDBlack Thanks for replying, I did set that as well. Here is my config:
https://gist.github.com/kris1710/9e756539419e507b129e3ecbfabcb9bc
The out of memory comes if i call ```sparknet.fitMultiDataSet(loadedPaths)```
However, I tried this :
take a subset of loadedpaths using 
```
val smallLoadedPaths = loadedPaths.zipWithIndex.filter(_._2 < 10).map(_._1)
for(i <-1 to 100){
sparknet.fitPathsMultiDataSet(smallLoadedPaths)
System.gc()
}
```
It works, 
I looked at memory usage of a node without calling gc at every loop , and it kept increasing by additionally 5gb with every fit call to smallLoadedPaths. This means, If i have to fit to the loadedpaths , I cannot do that without explicitly calling gc? I had to downgrade the batchsizeperworker=4 in my trainingmaster to make  `sparknet.fitPathsMultiDataSet(smallLoadedPaths)`  work.  
```
val trainingMaster: ParameterAveragingTrainingMaster = new ParameterAveragingTrainingMaster.Builder(50,32).workerPrefetchNumBatches(0).averagingFrequency(4).batchSizePerWorker(4).build()	
```
How to usually decide the batchsizeperworker, Isnt 4 too less? I mean the worker would update the weights( backprop) after every 4 observations and the averaging would be after `4*4=16` .
